### PR TITLE
Scheduler

### PR DESCRIPTION
**Pull request by CalvinFronda**

### CODE REVIEW

File: src/components/FeatureViews/Registration/RegistrationCart.js [hunk 1]
```diff
@@ -376,6 +376,8 @@ function RegistrationCart(props) {
                             new_course = {
                                 ...new_course,
                                 ...schedule,
+                                subject: "1:1 " + props.instructorAccounts[new_course.instructor].name +
```
Comment (NeilAllavarpu) [inline src/components/FeatureViews/Registration/RegistrationCart.js]: Optional: use [template literals](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Template_literals).

File: src/components/FeatureViews/Scheduler/SessionView.js [hunk 2]
```diff
@@ -1,15 +1,19 @@
...
 import BackButton from "../../BackButton.js";
 import SessionActions from "./SessionActions";
 import "../../../theme/theme.scss";
 import "./scheduler.scss";
-import {NavLink} from "react-router-dom";
+import { NavLink } from "react-router-dom";
 // import TimeSelector from "../../Form/TimeSelector";
 
+import * as calendarActions from "../../../actions/calendarActions";
+import * as courseActions from "../../../actions/apiActions";
```
Comment (NeilAllavarpu) [inline src/components/FeatureViews/Scheduler/SessionView.js]: Suggest renaming this to `apiActions` to make it clearer.

File: src/components/FeatureViews/Scheduler/SessionView.js [hunk 3]
```diff
@@ -31,36 +41,109 @@ class SessionView extends Component {
...
+                },
+                2: {
+                    name: "Joe Buddy",
+                    id: 4,
+                },
+                3: {
+                    name: "Able Strong Body",
+                    id: 11,
+                }
+            }
```
Comment (NeilAllavarpu) [inline src/components/FeatureViews/Scheduler/SessionView.js]: Remove default data from the component.

File: src/components/FeatureViews/Scheduler/SessionView.js [hunk 4]
```diff
@@ -31,36 +41,109 @@ class SessionView extends Component {
...
+                    courseData,
+                };
+            });
+
+        }
     }
 
+
+
+
```
Comment (NeilAllavarpu) [inline src/components/FeatureViews/Scheduler/SessionView.js]: Unnecessary whitespace.

File: src/components/FeatureViews/Scheduler/SessionView.js [hunk 5]
```diff
@@ -31,36 +41,109 @@ class SessionView extends Component {
...
+    }
+
+    componentDidUpdate(prevProps, prevState, snapshot) {
+        if (this.props !== prevProps &&
+            this.props.courseSessions.length !== 0 &&
+            Object.entries(this.props.courses).length !== 0) {
+            this.setState(() => {
+                const sessionData = this.props.courseSessions.find((session) => {
+                    return session.course === Number(this.props.match.params.course_id) &&
+                        session.id === Number(this.props.match.params.session_id);
```
Comment (NeilAllavarpu) [inline src/components/FeatureViews/Scheduler/SessionView.js]: Suggest converting both to Number before doing the comparison, or converting both to String.

File: src/actions/apiActions.js [hunk 6]
```diff
@@ -197,17 +197,21 @@ export const formatCourse = (formCourse, type) =>{
...
-        }
+        let Hour = time.substr(17, 2);
+        let to12HourTime = (Hour % 12) || 12;
+        let ampm = Hour < 12 ? " am" : " pm";
+        formattedTime = to12HourTime + time.substr(19, 3) + ampm;
     } else {
         formattedTime = time;
     }
-    return formattedTime.toTimeString().substring(0,5)
+    return formattedTime;
```
Comment (NeilAllavarpu) [inline src/actions/apiActions.js]: Move these to a separate `utils` file (there should be one under `src/` on the latest version of development.

File: src/actions/apiActions.js [hunk 7]
```diff
@@ -197,17 +197,21 @@ export const formatCourse = (formCourse, type) =>{
...
-        if(time.indexOf("AM") > -1 || time.indexOf("PM") > -1){
-            formattedTime = new Date();
-            formattedTime.setHours(Number(time.substring(0,time.indexOf(":"))));
-            formattedTime.setMinutes(Number(time.substring(time.indexOf(":")+1,time.indexOf(" "))));
-            formattedTime.setSeconds(0);
-        }
+        let Hour = time.substr(17, 2);
+        let to12HourTime = (Hour % 12) || 12;
+        let ampm = Hour < 12 ? " am" : " pm";
+        formattedTime = to12HourTime + time.substr(19, 3) + ampm;
```
Comment (NeilAllavarpu) [inline src/actions/apiActions.js]: Suggest using the [Date object](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date)'s methods for getting hours of the time instead of parsing a string, especially since you seem to be converting the created Data objects to strings before calling this function.

File: src/actions/apiActions.js [hunk 8]
```diff
@@ -197,17 +197,21 @@ export const formatCourse = (formCourse, type) =>{
     };
 }
 
-const parseTime = (time) =>{
+const courseName = (form, type) => {
+    if(type === "T"){
+        return "1:1 " + form["Instructor"].value + form[""]
+    }
```
Comment (NeilAllavarpu) [inline src/actions/apiActions.js]: Is there supposed to be a default or alternative return?

File: src/components/FeatureViews/Scheduler/SessionView.js [hunk 9]
```diff
@@ -31,36 +41,109 @@ class SessionView extends Component {
...
+            });
+
+        }
     }
 
+
+
+
     render() {
+        let instructor = this.state.courseData && this.props.instructors[this.state.courseData.instructor_id] ? this.props.instructors[this.state.courseData.instructor_id] : { name: "N/A" };
```
Comment (NeilAllavarpu) [inline src/components/FeatureViews/Scheduler/SessionView.js]: Break this up over a few lines for readability.

File: src/components/FeatureViews/Scheduler/SessionView.js [hunk 10]
```diff
@@ -31,36 +41,109 @@ class SessionView extends Component {
...
+        let dayConverter = {
```
Comment (NeilAllavarpu) [inline src/components/FeatureViews/Scheduler/SessionView.js]: Use `const` whenever the variable does not change.

Comment (NeilAllavarpu) [inline src/components/FeatureViews/Scheduler/SessionView.js]: This can be moved outside the component as a constant since it does not depend on state, props, or Redux.

File: src/components/FeatureViews/Scheduler/SessionView.js [hunk 11]
```diff
@@ -31,36 +41,109 @@ class SessionView extends Component {
...
+            0: "Sunday",
+            1: "Monday",
+            2: "Tuesday",
+            3: "Wednesday",
+            4: "Thursday",
+            5: "Friday",
+            6: "Saturday"
+        }
+        const styles = (username) => ({
```
Comment (NeilAllavarpu) [inline src/components/FeatureViews/Scheduler/SessionView.js]: Same as `dayConverter`

File: src/components/FeatureViews/Scheduler/SessionView.js [hunk 12]
```diff
@@ -71,54 +154,78 @@ class SessionView extends Component {
...
+                                <Typography varient="body1">{this.state.courseData && this.state.courseData.subject} </Typography>
                             </Grid>
                             <Grid
                                 item
-                                xs="auto">
-                                <Typography variant="h5"> Date & Time </Typography>
-                                <Typography variant="body1">{this.state.sessionData.start}</Typography>
+                                xs={6}>
+                                <Typography variant="h5"> Room</Typography>
+                                <Typography variant="body1">{this.state.courseData && (this.state.courseData.room_id || "TBA")}</Typography>
```
Comment (NeilAllavarpu) [inline src/components/FeatureViews/Scheduler/SessionView.js]: Parentheses may be misplaced; if `courseData` is null/undefined, the parenthesized OR is never reached, and nothing is shown.

File: src/components/Form/TimeSelector.js [hunk 13]
```diff
@@ -20,9 +20,9 @@ import CardActions from "@material-ui/core/CardActions";
 import Select from "@material-ui/core/es/Select/Select";
 import InputLabel from "@material-ui/core/InputLabel";
 
-Date.daysBetween = function( date1, date2 ) {
+Date.daysBetween = function (date1, date2) {
```
Comment (NeilAllavarpu) [inline src/components/Form/TimeSelector.js]: Try not to define functions on built-ins, instead make this a separate function.

File: src/components/Form/TimeSelector.js [hunk 14]
```diff
@@ -147,40 +147,40 @@ class TimeSelector extends Component {
         }
         // get instructor's current courses within this time
         try {
-            this.props.instructorID
-        } catch(error){
+            // this.props.instructorID
+        } catch (error) {
             console.error(error)
         }
```
Comment (NeilAllavarpu) [inline src/components/Form/TimeSelector.js]: Remove pointless try/catch

File: src/components/FeatureViews/Scheduler/ClassSessionView.js [hunk 15]
```diff
@@ -0,0 +1,85 @@
...
+            ...bindActionCreators(userActions, dispatch),
+            ...bindActionCreators(registrationActions, dispatch),
+        }),
+        [dispatch]
+    );
+
+    useEffect(() => {
+        setClassData(props.courseData)
+        // eslint-disable-next-line react-hooks/exhaustive-deps
+    }, [])
```
Comment (NeilAllavarpu) [inline src/components/FeatureViews/Scheduler/ClassSessionView.js]: Pointless to store props into a state variable - plus, the way it is currently written, this will never update when `props.courseData` is updated, since that is not in the dependencies. Only keep if you want only the original version of `courseData`

File: src/components/FeatureViews/Scheduler/Scheduler.js [hunk 16]
```diff
@@ -47,39 +52,80 @@ class Scheduler extends Component {
...
-        const newEvents = items.filter((item) => item.type === "C");
-        this.setState((prevState) => (
-            {
-                "calendarEvents": prevState.calendarEvents = newEvents,
-            }));
-        sessionStorage.setItem("calendarEvent", JSON.stringify(newEvents));
+    componentDidUpdate(prevProps, prevState, snapshot) {
+        if (this.props !== prevProps && this.props.sessions !== "" &&
+            !(Object.entries(this.props.courses).length === 0 &&
+                this.props.courses.constructor === Object) &&
```
Comment (NeilAllavarpu) [inline src/components/FeatureViews/Scheduler/Scheduler.js]: What exactly are you checking for here? You may not need to compare the constructor.

File: src/components/FeatureViews/Scheduler/Scheduler.js [hunk 17]
```diff
@@ -47,39 +52,80 @@ class Scheduler extends Component {
...
+            !(Object.entries(this.props.instructors).length === 0 &&
+                this.props.instructors === Object) &&
```
Comment (NeilAllavarpu) [inline src/components/FeatureViews/Scheduler/Scheduler.js]: Same as above.

File: src/components/FeatureViews/Scheduler/Scheduler.js [hunk 18]
```diff
@@ -141,15 +187,13 @@ class Scheduler extends Component {
...
+            let endTime = new Date(end).toTimeString();
             // Converts 24hr to 12 hr time
             function timeConverter(time) {
-
                 let Hour = time.substr(0, 2);
                 let to12HourTime = (Hour % 12) || 12;
-                let ampm = Hour < 12 ? "a" : "p";
+                let ampm = Hour < 12 ? " am" : " pm";
                 time = to12HourTime + time.substr(2, 3) + ampm;
                 return time;
```
Comment (NeilAllavarpu) [inline src/components/FeatureViews/Scheduler/Scheduler.js]: Seems similar to your `parseTime` function - consider reusing that.

File: src/components/FeatureViews/Scheduler/Scheduler.js [hunk 19]
```diff
@@ -200,38 +243,83 @@ class Scheduler extends Component {
...
-            "currentDate": date,
+        this.setState(() => {
+            this.props.calendarActions.fetchSessions({
+                config: {
+                    params: {
+                        time_frame: filter,
+                        view_option: this.viewOptions[this.state.calendarFilterValue],
+                    }
+                }
+            });
```
Comment (NeilAllavarpu) [inline src/components/FeatureViews/Scheduler/Scheduler.js]: Move this outside the `setState`, as it is not involved in setting the state. Also, filter does not appear to be initialized.

File: src/components/FeatureViews/Scheduler/Scheduler.js [hunk 20]
```diff
@@ -200,38 +243,83 @@ class Scheduler extends Component {
...
+                    time_shift: this.state.timeShift + 1,
+                }
+            }
+        });
+        this.setState((state) => {
+            return {
+                "currentDate": date,
+                "timeShift": state.timeShift + 1,
+            }
         });
```
Comment (NeilAllavarpu) [inline src/components/FeatureViews/Scheduler/Scheduler.js]: Can be shortened to:

    this.setState((state) => ({
        // object properties here
    }));

File: src/components/FeatureViews/Scheduler/Scheduler.js [hunk 21]
```diff
@@ -200,38 +243,83 @@ class Scheduler extends Component {
...
+                    view_option: this.viewOptions[this.state.calendarFilterValue],
+                    time_shift: 0,
+                }
+            }
+        });
         this.setState({
             "currentDate": date,
+            "timeShift": 0,
         });
     }
```
Comment (NeilAllavarpu) [inline src/components/FeatureViews/Scheduler/Scheduler.js]: Since these three functions are very similar in their code, suggest a broader `goTo` function that takes in a parameter indicating previous, next, or today and handles these three cases.

File: src/components/FeatureViews/Scheduler/Scheduler.js [hunk 22]
```diff
@@ -293,7 +381,8 @@ class Scheduler extends Component {
                 session["title"] = session.title;
                 session["description"] = session.description ? session.description : "";
                 session["type"] = this.props.courses[session.course_id].type;
-                session["resourceId"] = this.props.courses[session.course_id].room_id;
+                session["resourceId"] = this.props.courses[session.course_id].room_id ? this.props.courses[session.course_id].room_id : 1;
```
Comment (NeilAllavarpu) [inline src/components/FeatureViews/Scheduler/Scheduler.js]: Can be shortened to: `session["resourceId"] = this.props.courses[session.course_id].room_id || 1;`

File: src/components/FeatureViews/Scheduler/Scheduler.js [hunk 23]
```diff
@@ -314,16 +403,24 @@ class Scheduler extends Component {
 
     // This function is used in material-ui for the eventhandler
     handleFilterChange = (name) => (event) => {
-        this.setState({
-            [name]: event.target.value,
-        });
         if (event.target.value) {
-            const items = this.getEvents();
-            const newEvents = items.filter((item) => item.type === event.target.value);
+            const items = this.state.calendarEvents;
```
Comment (NeilAllavarpu) [inline src/components/FeatureViews/Scheduler/Scheduler.js]: Unused variable, can be removed.

File: src/components/FeatureViews/Scheduler/Scheduler.js [hunk 24]
```diff
@@ -368,132 +465,141 @@ class Scheduler extends Component {
...
-                                    <ViewListIcon />
-                                </IconButton>
-                            </Grid>
-                            <Grid item>
-                                <IconButton
-                                    className={"next-month"}
-                                    aria-label='next-month'>
-                                    <SearchIcon />
-                                </IconButton>
+                    <Grid conatiner className="scheduler-wrapper">
```
Comment (NeilAllavarpu) [inline src/components/FeatureViews/Scheduler/Scheduler.js]: `container` is misspelled.

File: src/components/FeatureViews/Scheduler/Scheduler.js [hunk 25]
```diff
@@ -368,132 +465,141 @@ class Scheduler extends Component {
...
-                                >Resource</Button>
-                            </Grid>
-                        </div>
-                    </div>
-                    <Grid item>
-                        <Grid className='omou-calendar'>
+                                </Grid>
+                                <Grid item>
+                                    <Button
+                                        onClick={() => { this.changeViewToCalendar(); }}
```
Comment (NeilAllavarpu) [inline src/components/FeatureViews/Scheduler/Scheduler.js]: Simplify to: onClick={this.changeViewToCalendar}

File: src/components/FeatureViews/Scheduler/Scheduler.js [hunk 26]
```diff
@@ -368,132 +465,141 @@ class Scheduler extends Component {
...
+                                    >Course</Button>
+                                </Grid>
+                                <Grid item >
+                                    <Button
+                                        onClick={() => { this.changeViewToResource(); }}
```
Comment (NeilAllavarpu) [inline src/components/FeatureViews/Scheduler/Scheduler.js]: Same as above.

File: src/components/FeatureViews/Scheduler/Scheduler.js [hunk 27]
```diff
@@ -514,6 +620,7 @@ class Scheduler extends Component {
                                 }}
                                 timeZone={"local"}
                                 eventMouseEnter={(this.state.resourceIcon) ? null : this.handleToolTip}
+                                eventClick={this.goToSessionView()}
```
Comment (NeilAllavarpu) [inline src/components/FeatureViews/Scheduler/Scheduler.js]: Similar to above.

File: src/components/FeatureViews/Scheduler/Scheduler.js [hunk 28]
```diff
@@ -514,6 +620,7 @@ class Scheduler extends Component {
                                 }}
                                 timeZone={"local"}
                                 eventMouseEnter={(this.state.resourceIcon) ? null : this.handleToolTip}
```
Comment (NeilAllavarpu) [inline src/components/FeatureViews/Scheduler/Scheduler.js]: Optional: Can be simplified to `!this.state.resourceIcon && this.handleToolTip`

File: src/components/FeatureViews/Scheduler/Scheduler.js (hunk 24 above)
Comment (dhuangg) [reply src/components/FeatureViews/Scheduler/Scheduler.js]: how do you catch stuff like this. mind = blown
